### PR TITLE
Add np.complex64 support in PL-Lightning

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
32-bit floats (`np.complex64`) are usable in PennyLane-Lightning.

**Description of the Change:**
This PR provides support for using 32-bit floats and 64-bit complex in Lightning. The following methods are updated: `apply`, `apply_lightning`, `adjoint_jacobian`, `_serialize_obs`, and `_serialize_ops` taking advantage of Complex64 C++ functions.  

**Benefits:**
This will reduce the memory usage to increase the performance of PL-Lightning methods.   

**Possible Drawbacks:**
As `dev.C_DTYPE` is set to `np.complex128` by default, to trigger routines based on 64-bit complex, one must ensure that `dev._state.dtype == np.complex64` for the given state-vector.  (A follow-up PR will address `np.complex64` support in PL)

### CODE REVIEW

File: pennylane_lightning/_serialize.py [hunk 1]
```diff
@@ -17,13 +17,24 @@
 from typing import List, Tuple
 
 import numpy as np
-from pennylane import BasisState, Hadamard, Projector, QubitStateVector, Rot
+from pennylane import (
```
Comment (mlxd) [inline pennylane_lightning/_serialize.py:20]: Good idea on this

File: tests/test_apply.py [hunk 2]
```diff
@@ -418,19 +424,19 @@ class TestExpval:
...
+            # (qml.PauliY, [1, 0], 0),
+            # (qml.PauliZ, [1, 0], 1),
+            # (qml.PauliZ, [0, 1], -1),
+            # (qml.PauliZ, [1 / math.sqrt(2), 1 / math.sqrt(2)], 0),
+            # (qml.Hadamard, [1, 0], 1 / math.sqrt(2)),
+            # (qml.Hadamard, [0, 1], -1 / math.sqrt(2)),
+            # (qml.Hadamard, [1 / math.sqrt(2), 1 / math.sqrt(2)], 1 / math.sqrt(2)),
+            # (qml.Identity, [1, 0], 1),
+            # (qml.Identity, [0, 1], 1),
+            # (qml.Identity, [1 / math.sqrt(2), -1 / math.sqrt(2)], 1),
```
Comment (mlxd) [inline tests/test_apply.py:439]: Are these failing?

File: tests/test_expval.py [hunk 3]
```diff
@@ -73,8 +73,10 @@ def test_paulix_expectation(self, theta, phi, qubit_device_3_wires, tol):
             rotations=[*O1.diagonalizing_gates(), *O2.diagonalizing_gates()],
         )
 
-        res = np.array([dev.expval(O1), dev.expval(O2)])
-        assert np.allclose(res, np.array([np.sin(theta) * np.sin(phi), np.sin(phi)]), tol)
+        res = np.array([dev.expval(O1), dev.expval(O2)], dtype=np.complex64)
```
Comment (mlxd) [inline tests/test_expval.py:76]: Is this co-opting a `complex128` test with `complex64` data? If so, I think we should keep the original and add a new one.

File: tests/test_serialize.py [hunk 4]
```diff
@@ -123,93 +134,112 @@ def test_tensor_non_tensor_return(self, monkeypatch):
...
         """Test for a comprehensive range of returns"""
         wires_dict = {"a": 0, 1: 1, "b": 2, -1: 3, 3.141: 4, "five": 5, 6: 6, 77: 7, 9: 8}
-        I = np.eye(2).astype(np.complex128)
-        X = qml.PauliX.matrix.astype(np.complex128)
-        Y = qml.PauliY.matrix.astype(np.complex128)
-        Z = qml.PauliZ.matrix.astype(np.complex128)
+        I = np.eye(2).astype(np.complex64)
+        X = qml.PauliX.matrix.astype(np.complex64)
+        Y = qml.PauliY.matrix.astype(np.complex64)
+        Z = qml.PauliZ.matrix.astype(np.complex64)
```
Comment (mlxd) [inline tests/test_serialize.py:236]: Should we have both?

File: tests/test_serialize.py [hunk 5]
```diff
@@ -381,7 +417,7 @@ def test_integration(self):
                     [],
                     [],
                     [],
-                    qml.QubitUnitary(np.eye(4, dtype=np.complex128), wires=[0, 1]).matrix,
+                    qml.QubitUnitary(np.eye(4, dtype=np.complex64), wires=[0, 1]).matrix,
```
Comment (mlxd) [inline tests/test_serialize.py:420]: Should the type be defined outside, or should we have tests for both?

**Review by mlxd — CHANGES_REQUESTED** [review]
Nice work on this @maliasadi 
I have a few questions:

File: tests/test_adjoint_jacobian.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 Tests for the ``adjoint_jacobian`` method of LightningQubit.
 """
 import math
+from numpy.core.arrayprint import dtype_short_repr
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply tests/test_adjoint_jacobian.py:18]: It's added automatically by ??. Removed!

File: tests/test_adjoint_jacobian.py [hunk 7]
```diff
@@ -161,9 +162,13 @@ def test_unsupported_hermitian_expectation(self, dev):
...
-    def test_pauli_rotation_gradient(self, G, theta, tol, dev):
+    @pytest.mark.parametrize("C", [np.complex64, np.complex128])
+    def test_pauli_rotation_gradient(self, G, theta, tol, dev, C):
         """Tests that the automatic gradients of Pauli rotations are correct."""
 
+        # Test adjoint_jacobian for type C
+        dev._state = dev._asarray(dev._state, C)
+
         with qml.tape.JacobianTape() as tape:
             qml.QubitStateVector(np.array([1.0, -1.0]) / np.sqrt(2), wires=0)
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply tests/test_adjoint_jacobian.py:173]: Yes! We don't need to cast within the tape. This is due to the fact that PL supports any complex types and neither update nor change the dtype of a given `_state`.

File: tests/test_adjoint_jacobian.py [hunk 8]
```diff
@@ -504,8 +541,8 @@ def f(params1, params2):
             qml.RY(tf.cos(params2), wires=[0])
             return qml.expval(qml.PauliZ(0))
 
-        params1 = tf.Variable(0.3, dtype=tf.float64)
-        params2 = tf.Variable(0.4, dtype=tf.float64)
+        params1 = tf.Variable(0.3, dtype=tf.float128)
+        params2 = tf.Variable(0.4, dtype=tf.float128)
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply tests/test_adjoint_jacobian.py:545]: You're totally right! My bad :slightly_smiling_face:

File: tests/test_serialize.py (hunk 4 above)
Comment (maliasadi) [reply tests/test_serialize.py:236]: Of course! I should have added already :slightly_smiling_face:

File: tests/test_apply.py (hunk 2 above)
Comment (maliasadi) [reply tests/test_apply.py:439]: Good catch! No. I just forgot to uncomment them after debugging.

**Comment by maliasadi** [comment]
## Performance 
To compare the performance of `.complex64` (C64) and `.complex128` (C128) in both PL and PL-Lightning, `default.qubit` and `lightning.qubit` are used to compute `adjoint_jacobian` for a set of tapes as follows. Note. the results are the median of 20 tries in [ms].

1. **tape_Rx**: 
```Python 
	params = np.linspace(-2 * np.pi, 2 * np.pi, wires)
	with qml.tape.JacobianTape() as tape:
		for idx in range(wires):
			qml.RX(params[idx], wires=idx)
		for idx in range(wires):
			qml.expval(qml.PauliZ(idx))
```
wires | default.qubit-C64 | default.qubit-C128 | lightning.qubit-C64 | lightning.qubit-C128
-- | -- | -- | -- | --
2 | 5.449665 | 5.439571 | 0.581803 | 0.582741
4 | 14.360847 | 14.430185 | 0.909154 | 0.909324
6 | 27.448523 | 27.297865 | 1.252368 | 1.252175
8 | 44.884279 | 44.709587 | 1.678570 | 1.656784
10 | 69.201878 | 69.155518 | 2.407422 | 2.365377
12 | 112.271834 | 111.883858 | 5.076586 | 4.267965
14 | 258.936116 | 257.424846 | 13.646876 | 12.163511
16 | 1429.513199 | 1412.968571 | 50.946691 | 46.829497
18 | 3566.677459 | 3543.637094 | 251.263023 | 296.666509
20 | 14794.269916 | 15006.239704 | 1440.236383 | 1501.908428

2. **tape_Rot**
```Python
	params = np.array([np.pi, np.pi ** 3, np.sqrt(2) * np.pi])
	with qml.tape.JacobianTape() as tape:
		qml.QubitStateVector(np.array([1.0, -1.0]) / np.sqrt(2), wires=0)
		for idx in range(wires):
			qml.Rot(*params, wires=[idx])
		for idx in range(wires):
			qml.expval(qml.PauliZ(idx))
```
nwires | default.qubit-C64 | default.qubit-C128 | lightning.qubit-C64 | lightning.qubit-C128
-- | -- | -- | -- | --
2 | 9.986461 | 9.966027 | 85.396899 | 84.907059
4 | 27.476278 | 27.404587 | 166.013697 | 160.209827
6 | 53.118247 | 53.495359 | 238.626306 | 235.287743
8 | 88.300125 | 87.645690 | 317.621658 | 316.910742
10 | 141.164253 | 138.494351 | 400.519065 | 399.066901
12 | 240.383712 | 237.925654 | 460.675714 | 458.162285
14 | 547.667520 | 534.426975 | 26.779808 | 25.193434
16 | 3766.410550 | 3876.882604 | 84.795516 | 92.952907
18 | 11048.476035 | 11008.885168 | 544.018366 | 693.615583
20 | 42432.884589 | 42988.092447 | 2173.447262 | 3033.272855

3. **tape_H**
```Python 
	with qml.tape.JacobianTape() as tape:
		for idx in range(wires-1):
			qml.Hadamard(wires=idx)
			qml.RX(0.543, wires=idx)
			qml.CNOT(wires=[idx, idx+1])
			qml.Rot(1.3, -2.3, 0.5, wires=[idx])
			qml.RZ(-0.5, wires=idx)
			qml.RY(0.5, wires=idx+1).inv()
			qml.CNOT(wires=[idx, idx+1])
		qml.expval(qml.PauliX(wires=0))
		qml.expval(qml.PauliZ(wires=wires-1))
```
nwires | default.qubit-C64 | default.qubit-C128 | lightning.qubit-C64 | lightning.qubit-C128
-- | -- | -- | -- | --
2 | 5.078271 | 5.092969 | 0.866478 | 0.861768
4 | 14.875412 | 14.862901 | 1.395402 | 1.365034
6 | 24.930872 | 24.888357 | 1.902398 | 1.902211
8 | 36.747390 | 36.483556 | 2.722451 | 2.662496
10 | 53.173519 | 52.868434 | 4.200512 | 4.550193
12 | 89.301016 | 89.616729 | 8.985506 | 9.441009
14 | 211.832398 | 212.154510 | 31.472651 | 33.722760
16 | 1403.903258 | 1595.419812 | 155.698221 | 161.448386
18 | 5223.126031 | 5231.572310 | 723.134491 | 812.019996
20 | 18458.793251 | 18338.228668 | 3562.543442 | 4223.660734

**Review by mlxd — APPROVED** [review]
Hi @maliasadi nice work on putting this together.

As discussed elsewhere regarding the performance, I think this is fine to go in as is (once the 2 comments are resolved). With an improvement to the kernels themselves, the performance gains should hopefully come with an improvement to vectorization and SIMD usage.

File: pennylane_lightning/lightning_qubit.py [hunk 9]
```diff
@@ -38,15 +38,15 @@
...
         from lightning_qubit_ops import (
-            apply,
             StateVectorC64,
+            AdjointJacobianC64,
             StateVectorC128,
             AdjointJacobianC128,
         )
     else:
         from .lightning_qubit_ops import (
-            apply,
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply pennylane_lightning/lightning_qubit.py:50]: Thanks for the clarification :slightly_smiling_face:

File: tests/test_adjoint_jacobian.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply tests/test_adjoint_jacobian.py:185]: I get it! Types of the params converted to float32 in `_serialize.py` in Python level if they don't match the dtype of the given state.

**Pull request merged**